The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14154968095/job/39653050691) [(plain)](https://github.com/rust-lang/rust/commit/599a94d7fd3dab722907f25ac0746756fc5aba6e/checks/39653050691/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.9s done
#19 DONE 25.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
  319 |           fprintf (f, prefix);
      |           ~~~~~~~~^~~~~~~~~~~
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc: In function ‘bool find_and_merge_options(int, off_t, const char*, vec<cl_decoded_option>, bool, vec<cl_decoded_option>*, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc:1165:8: warning: ignoring return value of ‘ssize_t read(int, void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 1165 |   read (fd, data, length);
      |   ~~~~~^~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
failures:

---- [codegen] tests/codegen/slice-last-elements-optimization.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-last-elements-optimization/slice-last-elements-optimization.ll" "/checkout/tests/codegen/slice-last-elements-optimization.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen/slice-last-elements-optimization.rs:18:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: slice_start_index_len_fail
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-last-elements-optimization/slice-last-elements-optimization.ll:27:28: note: found here
; call core::slice::index::slice_start_index_len_fail
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/tests/codegen/slice-last-elements-optimization.rs:19:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: unreachable
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-last-elements-optimization/slice-last-elements-optimization.ll:29:2: note: found here
 unreachable
 ^~~~~~~~~~~

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-last-elements-optimization/slice-last-elements-optimization.ll
Check file: /checkout/tests/codegen/slice-last-elements-optimization.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
        1: ; ModuleID = 'slice_last_elements_optimization.a38ed2c75d2fac3b-cgu.0' 
        2: source_filename = "slice_last_elements_optimization.a38ed2c75d2fac3b-cgu.0" 
        3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
        4: target triple = "x86_64-unknown-linux-gnu" 
        5:  
        6: @alloc_027431e5eb338ea39b017bf5fd7b8e2c = private unnamed_addr constant [59 x i8] c"/checkout/tests/codegen/slice-last-elements-optimization.rs", align 1 
        7: @alloc_6d459e27c6cc82129b9dd3e69fc4941b = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_027431e5eb338ea39b017bf5fd7b8e2c, [16 x i8] c";\00\00\00\00\00\00\00\07\00\00\00\15\00\00\00" }>, align 8 
        8: @alloc_588610a193162d8682d35311392f0db9 = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_027431e5eb338ea39b017bf5fd7b8e2c, [16 x i8] c";\00\00\00\00\00\00\00\07\00\00\00'\00\00\00" }>, align 8 
        9:  
       10: ; Function Attrs: nonlazybind uwtable 
       11: define { ptr, i64 } @last_four_initial(ptr noalias noundef nonnull readonly align 1 %s.0, i64 noundef %s.1) unnamed_addr #0 { 
       12: start: 
       13:  %_3 = icmp ult i64 %s.1, 5 
       14:  %0 = add i64 %s.1, -4 
       15:  %start1.sroa.0.0 = select i1 %_3, i64 0, i64 %0 
       16:  %_6 = icmp ugt i64 %start1.sroa.0.0, %s.1 
       17:  br i1 %_6, label %bb4, label %bb5, !prof !3 
       18:  
       19: bb5: ; preds = %start 
       20:  %_10 = sub nuw i64 %s.1, %start1.sroa.0.0 
       21:  %_11 = getelementptr inbounds i8, ptr %s.0, i64 %start1.sroa.0.0 
       22:  %1 = insertvalue { ptr, i64 } poison, ptr %_11, 0 
       23:  %2 = insertvalue { ptr, i64 } %1, i64 %_10, 1 
       24:  ret { ptr, i64 } %2 
       25:  
       26: bb4: ; preds = %start 
       27: ; call core::slice::index::slice_start_index_len_fail 
not:18                                !~~~~~~~~~~~~~~~~~~~~~~~~~  error: no match expected
       28:  tail call void @_RNvNtNtCs4SxWbCcFbBg_4core5slice5index26slice_start_index_len_fail(i64 noundef %start1.sroa.0.0, i64 noundef %s.1, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_6d459e27c6cc82129b9dd3e69fc4941b) #3 
       29:  unreachable 
not:19      !~~~~~~~~~~  error: no match expected
       30: } 
       31:  
       32: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
       33: define { ptr, i64 } @last_four_optimized(ptr noalias noundef nonnull readonly align 1 %s.0, i64 noundef %s.1) unnamed_addr #1 { 
       34: start: 
       35:  %_2 = icmp ult i64 %s.1, 5 
       36:  %0 = getelementptr i8, ptr %s.0, i64 %s.1 
       37:  %_17 = getelementptr i8, ptr %0, i64 -4 
       38:  %_0.sroa.3.0 = select i1 %_2, i64 %s.1, i64 4 
       39:  %_0.sroa.0.0 = select i1 %_2, ptr %s.0, ptr %_17 
       40:  %1 = insertvalue { ptr, i64 } poison, ptr %_0.sroa.0.0, 0 
       41:  %2 = insertvalue { ptr, i64 } %1, i64 %_0.sroa.3.0, 1 
       42:  ret { ptr, i64 } %2 
       43: } 
       44:  
       45: ; Function Attrs: nonlazybind uwtable 
       46: define { ptr, i64 } @test_bounds_check_happens(ptr noalias noundef nonnull readonly align 1 %s.0, i64 noundef %s.1, i64 noundef %i) unnamed_addr #0 { 
       47: start: 
       48:  %_3 = icmp ugt i64 %i, %s.1 
       49:  br i1 %_3, label %bb1, label %bb2, !prof !3 
       50:  
       51: bb2: ; preds = %start 
       52:  %_8 = sub nuw i64 %s.1, %i 
       53:  %_9 = getelementptr inbounds i8, ptr %s.0, i64 %i 
       54:  %0 = insertvalue { ptr, i64 } poison, ptr %_9, 0 
       55:  %1 = insertvalue { ptr, i64 } %0, i64 %_8, 1 
       56:  ret { ptr, i64 } %1 
       57:  
       58: bb1: ; preds = %start 
       59: ; call core::slice::index::slice_start_index_len_fail 
       60:  tail call void @_RNvNtNtCs4SxWbCcFbBg_4core5slice5index26slice_start_index_len_fail(i64 noundef %i, i64 noundef %s.1, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_588610a193162d8682d35311392f0db9) #3 
       61:  unreachable 
       62: } 
       63:  
       64: ; core::slice::index::slice_start_index_len_fail 
       65: ; Function Attrs: cold noinline noreturn nonlazybind uwtable 
       66: declare void @_RNvNtNtCs4SxWbCcFbBg_4core5slice5index26slice_start_index_len_fail(i64 noundef, i64 noundef, ptr noalias noundef readonly align 8 dereferenceable(24)) unnamed_addr #2 
       67:  
       68: attributes #0 = { nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
       69: attributes #1 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
       70: attributes #2 = { cold noinline noreturn nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
       71: attributes #3 = { noreturn } 
       72:  
       73: !llvm.module.flags = !{!0, !1} 
       74: !llvm.ident = !{!2} 
       75:  
       76: !0 = !{i32 8, !"PIC Level", i32 2} 
       77: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
       78: !2 = !{!"rustc version 1.88.0-nightly (66b6872a3 2025-03-30)"} 
       79: !3 = !{!"branch_weights", i32 1, i32 2000} 
>>>>>>
------------------------------------------



```

</details>

--- a/tests/codegen/slice-last-elements-optimization.rs
+++ b/tests/codegen/slice-last-elements-optimization.rs
@@ -1,0 +1,37 @@
+//@ compile-flags: -Copt-level=3
+//@ only-x86_64
+//@ min-llvm-version: 20
+#![crate_type = "lib"]
+
+// This test verifies that LLVM 20 properly optimizes the bounds check
+// when accessing the last few elements of a slice with proper conditions.
+// Previously, this would generate an unreachable branch to
+// slice_start_index_len_fail even when the bounds check was provably safe.
+
+// CHECK-LABEL: @last_four_initial(
+#[no_mangle]
+pub fn last_four_initial(s: &[u8]) -> &[u8] {
+    // Previously this would generate a branch to slice_start_index_len_fail
+    // that is unreachable. The LLVM 20 fix should eliminate this branch.
+    // CHECK-NOT: slice_start_index_len_fail
+    // CHECK-NOT: unreachable
+    let start = if s.len() <= 4 { 0 } else { s.len() - 4 };
+    &s[start..]
+}
+
+// CHECK-LABEL: @last_four_optimized(
+#[no_mangle]
+pub fn last_four_optimized(s: &[u8]) -> &[u8] {
+    // This version was already correctly optimized before the fix in LLVM 20.
+    // CHECK-NOT: slice_start_index_len_fail
+    // CHECK-NOT: unreachable
+    if s.len() <= 4 { &s[0..] } else { &s[s.len() - 4..] }
+}
+
+// Just to verify we're correctly checking for the right thing
+// CHECK-LABEL: @test_bounds_check_happens(
+#[no_mangle]
+pub fn test_bounds_check_happens(s: &[u8], i: usize) -> &[u8] {
+    // CHECK: slice_start_index_len_fail
+    &s[i..]
+}

